### DISCUSSION TITLE
What is correct method to do multi level queries

### DISCUSSION THREAD

**Discussion by kinekt4** [downforcetech/fluidxml.php#53] [Questions]
I want to do multi level iterations.

ie. I want to select all the rows (`tr`). Iterate over each row. Then select the columns (`td`).

I can't see an easy way to do it with the methods available.

eg. Given the following html doc:
```
<table id="data-table">
  <tbody>
    <tr>
      <td>Row 1</td>
      <td>Col 1</td>
      <td>Col 2</td>
    </tr>
    <tr>
      <td>Row 2</td>
      <td>Col 1</td>
      <td>Col 2</td>
    </tr>
  </tbody>
</table>
```
To select the rows I would:
```
$rows = $fXml->query('table#data-table tr');
```
Now I want to select the columns based on the `$rows` result

Ideally, I was thinking along the lines of:
```
$rows->each(function($i, $domNode) {
    $cols = $this->query('td');
});
```
But I know that `$this` in the closure is a `FluidContext` object and doesn't have the `query` method.

Is there any other way to do it that I'm not aware of?

**Comment by daniele-orlando** [comment]
You have different options depending on the exact task you have to accomplish.

To iterate all rows and all columns, in batch, you can:
```php
$rows = $fxml->query('//tr');
$cols = $fxml->query('//td');

$rows->each(function ($idx) {
});
$cols->each(function ($idx) {
});
```

To do a depth traversal (a row first, then its columns), you can:
```php
$fxml->query('//tr')->each(function ($rowIdx) {
        $this->query('./td')->each(function ($colIdx) {
        });
});
```


> But I know that $this in the closure is a FluidContext object and doesn't have the query method.

Exactly because it is a FluidContext, you have the `->query()` method.

**Reply by kinekt4** [reply]
I didn't know you could do this:

```
$fxml->query('//tr')->each(function ($rowIdx) {
    $this->query('./td')->each(function ($colIdx) {
    });
});
```
For some reason I didn't think FluidContext had the `query()` method. I can now see it's implementing the `FluidInterface` which does indeed have the `query()` method.

I was doing this instead, which is not as efficient:
```
$fxml->query('//tr')->each(function ($rowIdx, $rDomNode) {

    $fxml = new FluidXml($rDomNode);
    $fxml ->query('./td')->each(function ($colIdx) {
        ...
    });
});
```
